### PR TITLE
Combobox base

### PR DESCRIPTION
**Pull request by nielsr98**
Added most base features of combobox and panel. By default this combobox will always trigger on click. More unit tests are required. Only basic unit tests are implemented thus far.

### CODE REVIEW

File: src/cdk-experimental/combobox/combobox.ts [hunk 1]
```diff
@@ -6,15 +6,192 @@
...
+
 
 @Directive({
   selector: '[cdkCombobox]',
   exportAs: 'cdkCombobox',
   host: {
-    'role': 'combobox'
+    'role': 'combobox',
+    '(click)': 'toggle()',
+    '[attr.aria-disabled]': 'disabled'
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:42]: The `aria-haspopup` and `aria-controls` should go here. 

I also realized just now that there's a mode of interaction where this element controls a listbox while focus remains on the combobox element and it uses `aria-activedescendant` to indicate which listbox option is active. Seems like something to consider later on

File: src/cdk-experimental/combobox/combobox.ts [hunk 2]
```diff
@@ -6,15 +6,192 @@
...
-export class CdkCombobox<T = unknown> {
+export class CdkCombobox<T = unknown> implements OnDestroy, AfterContentInit {
+  @Input('triggerFor')
+  get comboboxPanel(): CdkComboboxPanel<T> | undefined {
+    return this._comboboxPanel;
+  }
+  set comboboxPanel(panel: CdkComboboxPanel<T> | undefined) {
+    this._comboboxPanel = panel;
+  }
+  private _comboboxPanel: CdkComboboxPanel<T> | undefined;
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:53]:
```suggestion
  @Input('cdkComboboxTriggerFor')
  get panel(): CdkComboboxPanel<T> | undefined {
    return this._panel;
  }
  set panel(panel: CdkComboboxPanel<T> | undefined) {
    this._panel= panel;
  }
  private _panel: CdkComboboxPanel<T> | undefined;
```

File: src/cdk-experimental/combobox/combobox.ts [hunk 3]
```diff
@@ -6,15 +6,192 @@
...
+
+  @Input()
+  get value(): T {
+    return this._value;
+  }
+  set value(val: T) {
+    this._value = val;
+  }
+  private _value: T;
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:62]:
```suggestion
  @Input()
  value: T;
```
I'd avoid adding the getter/setter until you need it

File: src/cdk-experimental/combobox/combobox.ts [hunk 4]
```diff
@@ -6,15 +6,192 @@
...
+
+  @Input()
+  get disabled(): boolean {
+    return this._disabled;
+  }
+  set disabled(value: boolean) {
+    this._disabled = coerceBooleanProperty(value);
+  }
+  private _disabled: boolean = false;
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:71]: optional: you can condense simple getter/setters like this
```suggestion
  get disabled(): boolean { return this._disabled; }
  set disabled(value: boolean) { this._disabled = coerceBooleanProperty(value); }
  private _disabled: boolean = false;
```

File: src/cdk-experimental/combobox/combobox.ts [hunk 5]
```diff
@@ -6,15 +6,192 @@
...
+  private _getOverlayPositions(): ConnectedPosition[] {
+    return [
+      {originX: 'start', originY: 'bottom', overlayX: 'start', overlayY: 'top'},
+      {originX: 'start', originY: 'top', overlayX: 'start', overlayY: 'bottom'},
+      {originX: 'end', originY: 'bottom', overlayX: 'end', overlayY: 'top'},
+      {originX: 'end', originY: 'top', overlayX: 'end', overlayY: 'bottom'},
+    ];
+  }
+
+  private _getPortal() {
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:171]:
```suggestion
  private _getPanelContent() {
```

File: src/cdk-experimental/combobox/combobox-panel.ts [hunk 6]
```diff
@@ -6,12 +6,35 @@
...
 
-import {Directive} from '@angular/core';
+import {Directive, TemplateRef} from '@angular/core';
+import {Subject} from 'rxjs';
 
 @Directive({
   selector: 'ng-template[cdkComboboxPanel]',
   exportAs: 'cdkComboboxPanel',
+  host: {
+    'aria-controls': 'contentId',
```
[the remark this answers is not in the document]
Comment (nielsr98) [reply src/cdk-experimental/combobox/combobox-panel.ts:16]: So I guess using a Subject for these two as well would be appropriate? Then on register of the content the panel updated the Subject and the trigger can grab the new value.

File: src/cdk-experimental/combobox/combobox-panel.ts [hunk 7]
```diff
@@ -6,12 +6,35 @@
...
+  constructor(readonly _templateRef: TemplateRef<unknown>) {}
+
+  closePanel(data?: T) {
+    this.valueUpdated.next(data);
+  }
+
+  _registerContent(contentId: string, contentType: string) {
+    this.contentId = contentId;
+    if (contentType !== 'listbox' && contentType !== 'dialog') {
+      throw Error('CdkComboboxPanel content must be either a listbox or dialog');
```
[the remark this answers is not in the document]
Comment (nielsr98) [reply src/cdk-experimental/combobox/combobox-panel.ts:36]: Gotcha will do.

File: src/cdk-experimental/combobox/combobox.ts (hunk 3 above)
Comment (nielsr98) [reply src/cdk-experimental/combobox/combobox.ts:62]: Ah okay.

File: src/cdk-experimental/combobox/combobox.ts [hunk 8]
```diff
@@ -6,15 +6,194 @@
...
+    const inputArray = coerceArray(input);
+    for (const action of inputArray) {
+      if (action !== 'focus' && action !== 'click' && action !== 'downKey' && action !== 'toggle') {
+        throw Error('Not a valid open action.');
+      } else {
+        actions.push(action);
+      }
+    }
+
+    return actions;
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:194]: I would just have this whole function be:
```suggestion
return input;
```
And add the rest of the behavior when you add support for splitting a string

File: src/cdk-experimental/combobox/combobox.ts [hunk 9]
```diff
@@ -6,15 +6,194 @@
...
 @Directive({
   selector: '[cdkCombobox]',
   exportAs: 'cdkCombobox',
   host: {
-    'role': 'combobox'
+    'role': 'combobox',
+    '(click)': 'toggle()',
+    '[attr.aria-disabled]': 'disabled',
+    '[attr.aria-controls]': 'contentId',
+    '[attr.aria-haspopup]': 'contentType'
```
Comment (jelbourn) [inline src/cdk-experimental/combobox/combobox.ts:43]: Add unit tests for these attributes?

**Pull request merged**